### ISSUE TITLE
Allow grid_sample to accept pixel units (absolute coordinates)

### ISSUE THREAD

**Issue by bnehoran** [pytorch/pytorch#36107]
_This issue is expanded from #24870 for reference and for additional discussions on implementation details. See that issue for context._

Add an option for the grid/flow input to `grid_sample` to be expressed in pixel units (absolute coordinates). This is much more natural for some applications, including image registration, where the grid-like input is actually a displacement field, and this eliminates the need for the user to convert to normalized [-1,1] coordinates, which is prone to user error, (especially when users don’t know which setting of `align_corners` to target).

This would help simplify the implementation of convolutional networks to do Optical Flow Estimation and Image Registration tasks.

**Comment by bnehoran** [comment]
I've started working on this, but one place where I could use help is in deciding what to call the option.
Right now, I have a boolean option named `pixel_coords` as a proposed candidate. 
Of course, feel free to make a suggestion if you have another idea.

**Comment by tvercaut** [comment]
As detailed in #81868, I am encountering precision issues with `grid_sample` in `bilinear` mode at discrete pixel locations. Simple experimentation seems to point towards the fact that the current normalized coordinates induces this loss of precision. Having an option to bypass grid normalization and work in absolute pixel coordinates would thus also likely be helpful in terms of improving the numerical precision of the interpolation.

**Comment by vboussot** [comment]
Still wanted. Here is the precision cost, measured, from a use case where tiling is not optional.

We resample volumes patch by patch: the volume is walked in windows and each window is sampled from the
source with `grid_sample`. The grid is normalised by the extent of the tensor passed in, so the same
physical point sampled from a window and from the whole volume does not come back as the same value.

**Round trip of a coordinate through `(2i + 1) / n - 1` in float32:**

| extent | max error |
| --- | --- |
| 64 | 2.9e-06 voxel |
| 512 | 2.3e-05 |
| 1000 | 4.6e-05 |
| 3000 | 1.8e-04 |

```python
import torch

def norm(c, n): return (2.0 * c + 1.0) / n - 1.0
def unnorm(u, n): return ((u + 1.0) * n - 1.0) / 2.0

for n in (64, 512, 1000, 3000):
    c = torch.rand(200_000, dtype=torch.float64) * (n - 2) + 1.0
    print(n, (unnorm(norm(c.float(), n), n).double() - c).abs().max().item())
```

**What that does to a tiled read.** Sample the same physical points out of the whole volume and out of a
window of it. The coordinate is chosen so it is exactly representable in float32 in both frames, so the
only difference between the two calls is the denominator:

| volume | window | offset | max diff | voxels differing |
| --- | --- | --- | --- | --- |
| 256 | 64 | 64 | 0 | 0.0% |
| 512 | 64 | 128 | 0 | 0.0% |
| 1024 | 128 | 256 | 0 | 0.0% |
| 300 | 64 | 100 | 1.97e-05 | 24.2% |
| 333 | 55 | 101 | 1.32e-05 | 57.8% |
| 500 | 91 | 137 | 6.57e-06 | 55.0% |
| 700 | 121 | 203 | 2.83e-05 | 57.3% |
| 999 | 100 | 300 | 5.87e-06 | 49.6% |
| 1500 | 96 | 500 | 5.00e-05 | 42.4% |
| 2000 | 100 | 700 | 6.50e-06 | 49.6% |

bilinear, `padding_mode="border"`, float32 payload in [0, 1), worst of 3 draws.

The first three rows are the control. When both extents are powers of two, `(2c + 1) / n` and its inverse
are exact scalings, both calls recover the same source index, and the outputs agree bit for bit: 84 of 84
configurations over three volume sizes, three window sizes, five offsets and two seeds. What decides the
divergence is whether the denominator divides exactly, not the payload and not the tiling as such.

```python
import torch, torch.nn.functional as F
def norm(c, n): return (2.0 * c + 1.0) / n - 1.0

N, ext, start = 700, 121, 203
vol = torch.rand(1, 1, N, N, N)
k = torch.randint(16, 16 * (ext - 3), (ext, ext, ext, 3))
pts = start + k.float() / 16.0                      # exact in float32 in both frames
assert torch.equal(pts - start, k.float() / 16.0)

whole = F.grid_sample(vol, norm(pts, N).unsqueeze(0), padding_mode="border", align_corners=False)
win = vol[:, :, start:start+ext, start:start+ext, start:start+ext].contiguous()
part = F.grid_sample(win, norm(pts - start, ext).unsqueeze(0), padding_mode="border", align_corners=False)

d = (whole - part).abs()
print(d.max().item(), (d > 0).float().mean().item())
```

The consequence is that how you tile changes the answer, which is the same effect @tvercaut reported here
in 2022 from the other direction.

In KonfAI (a patch based deep learning framework for medical imaging) we ended up not using `grid_sample`
for nearest and cubic at all, and doing the gather by hand over global voxel indices, precisely to keep a
streamed region and the whole volume bit identical. The coordinates are already voxel indices at the call
site; normalising them just to unnormalise them inside the kernel is where the bits go.

#41241 was closed as stale in 2022 with the CLA still outstanding. I have been working in `GridSampler`
recently (#194787) and am happy to open a PR for this covering CPU, CUDA, MPS, meta and the backward,
unless someone is already on it.
